types/react-select/src/Select.d.ts

```diff
@@ -103,7 +102,7 @@ export interface NamedProps<OptionType extends OptionTypeBase = { label: string;
   escapeClearsValue?: boolean;
   /* Custom method to filter whether an option should be displayed in the menu */
   filterOption?: ((
-    option: Option,
+    option: OptionType,
```

Comment:
I think you're conflating the `option` that is passed into `filterOption` and the `options` that the user passes in. The `option` that is passed into `filterOption` is `{ label: string; value: string; data: OptionType }` (see [the docs](https://react-select.com/props)). I made [my own attempt](https://github.com/Methuselah96/DefinitelyTyped/commit/79613ab03901d0698f3ce0f0610e1838ceaa584d) at what the types should be. Feel free to use it in this PR.

Reply:
I've updated the CustomFilterOption test example so it's hopefully more thorough

I think the generic argument of `<Select` ends up being of the `{ label: string; value: string; data: OptionType ` structure

Reply:
If the generic argument of `Select` ended up including a `data` property, wouldn't the user have to specify the `data` property when they provide the `options` to `Select`?

`filterOption` is used on [this line](https://github.com/JedWatson/react-select/blob/master/packages/react-select/src/Select.js#L1309). As you can see it is constructed on the spot and uses the `option` as the `data` property.

Reply:
Yeah I think I misread the types, I believe the generic argument is more loose:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/a3c43f54298a65e85ffe228200c3ec3f1345a9e8/types/react-select/src/stateManager.d.ts#L41

`OptionsTypeBase` is an alias for `{[key: string]: any}` so there aren't many restrictions on what is passed in and there is no requirement for a data property

Reply:
In the linked code for react-select, they pass in the `value` field as well which is usually `string`, but can be customized via the generic param.

And I think that's what allows the `filterOption` to work with the generic param:

```ts
function filterOption(option: GroupOption, rawInput: string): boolean {
  return (option.value.email + option.value.name).includes(rawInput);
}
```

Reply:
Based on the react-select documentation and Flow types, the value property is meant/expected to be a string. You can accomplish what you're trying to do by storing that data on a separate property like so:

```
interface GroupOption {
  label: string;
  value: string;
  metadata: {
    email: string
    name: string
  };
}

function filterOption(option: Option<GroupOption>, rawInput: string): boolean {
  return (option.data.metadata.email + option.data.metadata.name).includes(rawInput);
}

const values: GroupOption[] = [
    { label: "", value: "", metadata: { email: "", name: "" }}
];
```

Or if you're insistent on keeping the `value` as an object (which is not supported based on the documentation and Flow types) you can access it through the `data` property as well:

```
function filterOption(option: Option<GroupOption>, rawInput: string): boolean {
  return (option.data.value.email + option.data.value.name).includes(rawInput);
}
```

Reply:
Interesting, seems that the `OptionType` type param in `<Select` should be update to have a constraint along the lines of `{ label: string, value: string, data: any }`

Reply:
No, I think you're still misunderstanding the purpose of `OptionType`. `OptionType[]` (ignore grouped options) is the type of the variable that is passed into the `options` prop. It doesn't need to include `data`.

The `filterOption` prop takes a function that takes this as it's first parameter:
```
interface Option<OptionType extends OptionTypeBase> {
    label: string;
    value: string;
    data: OptionType;
}
```

Reply:
so data should be optional?

Also I've changed the example to use a separate `data` property instead of `value` for the object value

Reply:
Perhaps [this example](https://codesandbox.io/s/optimistic-monad-j967y?file=/src/App.js) will be helpful.

Here's the code:
![image](https://user-images.githubusercontent.com/693755/97348619-66760300-185c-11eb-8fce-7ed11a8ae63a.png)

Here's what gets logged to the console:
![image](https://user-images.githubusercontent.com/693755/97348573-53633300-185c-11eb-8ec8-418ea6747136.png)

The `values` variables is the type `OptionType[]`. The option parameter in `filterOption` is the type `{ label: string; value: string; data: OptionType }`.

The changes in [this commit](https://github.com/Methuselah96/DefinitelyTyped/commit/79613ab03901d0698f3ce0f0610e1838ceaa584d) are what I think it should be.